src/theory/bv/bvgauss.cpp

```diff
@@ -0,0 +1,692 @@
+      }
+    }
+  }
+
+  bool ispart = false;
+  for (size_t i = 0; i < nrows; ++i)
+  {
+    size_t pcol = i;
+    while (pcol < ncols && reslhs[i][pcol] == 0) ++pcol;
+    if (pcol >= ncols)
+    {
+      resrhs[i] = resrhs[i].euclidianDivideRemainder(prime);
+      if (resrhs[i] != 0) return BVGaussElim::Result::NONE; /* no solution */
+      continue;
+    }
+    for (size_t j = i; j < ncols; ++j)
+    {
+      if (reslhs[i][j] >= prime || reslhs[i][j] <= -prime)
+        reslhs[i][j] = reslhs[i][j].euclidianDivideRemainder(prime);
+      if (j > pcol && reslhs[i][j] != 0) ispart = true;
+    }
+  }
+  if (ispart) return BVGaussElim::Result::PARTIAL;
+
+  return BVGaussElim::Result::UNIQUE;
+}
+
+static bool is_bv_const(Node n)
+{
+  for (Kind k = n.getKind(); k != kind::CONST_BITVECTOR; k = n.getKind())
+  {
+    if (k == kind::BITVECTOR_CONCAT
+        && n[0] == utils::mkZero(utils::getSize(n[0])))
+    {
+      n = n[1];
+    }
+    else if (k == kind::BITVECTOR_ZERO_EXTEND)
+    {
+      n = n[0];
+    }
+    else
+    {
+      return false;
+    }
+  }
+  return n.getKind() == kind::CONST_BITVECTOR;
+}
+
+static Integer get_bv_const(Node n)
+{
+  Assert(is_bv_const(n));
+
+  for (Kind k = n.getKind(); k != kind::CONST_BITVECTOR; k = n.getKind())
+  {
+    if (k == kind::BITVECTOR_CONCAT)
+    {
+      Assert(n[0] == utils::mkZero(utils::getSize(n[0])));
+      n = n[1];
+    }
+    else
+    {
+      Assert(k == kind::BITVECTOR_ZERO_EXTEND);
+      n = n[0];
+    }
+  }
+  Assert(n.getKind() == kind::CONST_BITVECTOR);
+  return n.getConst<BitVector>().getValue();
+}
+
+BVGaussElim::Result BVGaussElim::gaussElimRewriteForUrem(
+    vector<Node> &equations, unordered_map<Node, Node, NodeHashFunction> &res)
+{
+  Node prime;
+  Integer iprime;
+  unordered_map<Node, vector<Integer>, NodeHashFunction> vars;
+  size_t neqs = equations.size();
+  vector<Integer> resrhs, rhs;
+  vector<vector<Integer>> reslhs,
+      lhs = vector<vector<Integer>>(neqs, vector<Integer>());
+
+  res = unordered_map<Node, Node, NodeHashFunction>();
+
+  for (size_t i = 0; i < neqs; ++i)
+  {
+    Node eq = equations[i];
+    Assert(eq.getKind() == kind::EQUAL);
+    Node urem, eqrhs;
+
+    if (eq[0].getKind() == kind::BITVECTOR_UREM)
+    {
+      urem = eq[0];
+      Assert(is_bv_const(eq[1]));
+      eqrhs = eq[1];
+    }
+    else
+    {
+      Assert(eq[1].getKind() == kind::BITVECTOR_UREM);
+      urem = eq[1];
+      Assert(is_bv_const(eq[0]));
+      eqrhs = eq[0];
+    }
+    if (getMinBwExpr(urem[0]) == 0)
+    {
+      Trace("bv-gauss-elim")
+          << "Minimum required bit-width exceeds given bit-width, "
+             "will not apply Gaussian Elimination."
+          << endl;
+      return BVGaussElim::Result::INVALID;
+    }
+    rhs.push_back(get_bv_const(eqrhs));
+
+    Assert(is_bv_const(urem[1]));
+    Assert(i == 0 || get_bv_const(urem[1]) == iprime);
+    if (i == 0)
+    {
+      prime = urem[1];
```

Comment:
This is intentional since I check that each equation is modulo the same `prime`. Passing `prime` as a parameter has no benefit since I have to untangle each equation anyhow (they are given as `(a1*b1 + ... + an*bn) % p = c`).

Reply:
Yeah, it is not super important, I just thought that it would save some code.